arm9/source/utils/nandcmac.c

```diff
@@ -196,6 +302,76 @@ u32 CalculateFileCmac(const char* path, u8* cmac) {
         u32 ret = FixAgbSaveCmac(agbsave, cmac, (cmac_type == CMAC_AGBSAVE) ? NULL : path);
         free(agbsave);
         return ret;
+    } else if (cmac_type == CMAC_AGBSAVE_EEPROM_512) {
+        AgbSaveHeader* agbsave = (AgbSaveHeader*) malloc(AGBSAVE_MAX_SIZE);
+        UINT br;
+        
+        if (!agbsave) return 1;
+        if ((fvx_qread(path, agbsave, 0x400, AGBSAVE_MAX_SIZE, &br) != FR_OK) || (br < 0x200) ||
+            (ValidateAgbSaveHeader(agbsave) != 0) || (0x200 + agbsave->save_size > br)) {
+            free(agbsave);
+            return 1;
+        }
+        
+        u32 ret = FixAgbSaveCmac(agbsave, cmac, path);
+        free(agbsave);
+        return ret;
+    } else if (cmac_type == CMAC_AGBSAVE_EEPROM_64K) {
+        AgbSaveHeader* agbsave = (AgbSaveHeader*) malloc(AGBSAVE_MAX_SIZE);
+        UINT br;
+        
+        if (!agbsave) return 1;
+        if ((fvx_qread(path, agbsave, 0x2200, AGBSAVE_MAX_SIZE, &br) != FR_OK) || (br < 0x200) ||
+            (ValidateAgbSaveHeader(agbsave) != 0) || (0x200 + agbsave->save_size > br)) {
+            free(agbsave);
+            return 1;
+        }
+        
+        u32 ret = FixAgbSaveCmac(agbsave, cmac, path);
+        free(agbsave);
+        return ret;
+    } else if (cmac_type == CMAC_AGBSAVE_SRAM_256K) {
+        AgbSaveHeader* agbsave = (AgbSaveHeader*) malloc(AGBSAVE_MAX_SIZE);
+        UINT br;
+        
+        if (!agbsave) return 1;
+        if ((fvx_qread(path, agbsave, 0x8200, AGBSAVE_MAX_SIZE, &br) != FR_OK) || (br < 0x200) ||
+            (ValidateAgbSaveHeader(agbsave) != 0) || (0x200 + agbsave->save_size > br)) {
+            free(agbsave);
+            return 1;
+        }
+        
+        u32 ret = FixAgbSaveCmac(agbsave, cmac, path);
+        free(agbsave);
+        return ret;
+    } else if (cmac_type == CMAC_AGBSAVE_FLASH_512K) {
+        AgbSaveHeader* agbsave = (AgbSaveHeader*) malloc(AGBSAVE_MAX_SIZE);
+        UINT br;
+        
+        if (!agbsave) return 1;
+        if ((fvx_qread(path, agbsave, 0x10200, AGBSAVE_MAX_SIZE, &br) != FR_OK) || (br < 0x200) ||
+            (ValidateAgbSaveHeader(agbsave) != 0) || (0x200 + agbsave->save_size > br)) {
+            free(agbsave);
+            return 1;
+        }
+        
+        u32 ret = FixAgbSaveCmac(agbsave, cmac, path);
+        free(agbsave);
+        return ret;
+    } else if (cmac_type == CMAC_AGBSAVE_FLASH_1M) {
+        AgbSaveHeader* agbsave = (AgbSaveHeader*) malloc(AGBSAVE_MAX_SIZE);
+        UINT br;
+        
+        if (!agbsave) return 1;
+        if ((fvx_qread(path, agbsave, 0x20200, AGBSAVE_MAX_SIZE, &br) != FR_OK) || (br < 0x200) ||
+            (ValidateAgbSaveHeader(agbsave) != 0) || (0x200 + agbsave->save_size > br)) {
+            free(agbsave);
+            return 1;
+        }
+        
+        u32 ret = FixAgbSaveCmac(agbsave, cmac, path);
+        free(agbsave);
+        return ret;
```

Comment:
Same issue here, all the code is exactly the same for each of the CMAC_AGBSAVE_X types, except for the qread offset parameter.